internal/service/ec2/transit_gateway_multicast_domain_test.go

```diff
@@ -0,0 +1,1007 @@
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("not found: %s", resourceName)
+		}
+
+		id := rs.Primary.ID
+		if id == "" {
+			return fmt.Errorf("no EC2 Transit Gateway Multicast Domain ID is set")
+		}
+
+		conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+		domain, err := tfec2.DescribeTransitGatewayMulticastDomain(conn, id)
+		if err != nil {
+			return err
+		}
+
+		if domain == nil {
+			return fmt.Errorf("EC2 Transit Gateway Multicast Domain (%s) not found", id)
+		}
+
+		state := aws.StringValue(domain.State)
+		if state != ec2.TransitGatewayMulticastDomainStateAvailable {
+			return fmt.Errorf(
+				"EC2 Transit Gateway Multicast Domain (%s) exists in non-available (%s) state", id, state)
+		}
+
+		*multicastDomain = *domain
+
+		return nil
+	}
+}
+
+func testAccCheckTransitGatewayMulticastDomainDisappears(domain *ec2.TransitGatewayMulticastDomain) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+		id := aws.StringValue(domain.TransitGatewayMulticastDomainId)
+		input := &ec2.DeleteTransitGatewayMulticastDomainInput{
+			TransitGatewayMulticastDomainId: aws.String(id),
+		}
+
+		log.Printf("[DEBUG] Deleting EC2 Transit Gateway Multicast Domain (%s): %s", id, input)
+		if _, err := conn.DeleteTransitGatewayMulticastDomain(input); err != nil {
+			return err
+		}
+
+		return tfec2.WaitForTransitGatewayMulticastDomainDeletion(conn, id)
+	}
+}
+
+func testAccCheckTransitGatewayMulticastDomainDestroy(s *terraform.State) error {
+	conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_ec2_transit_gateway_multicast_domain" {
+			continue
+		}
+
+		id := rs.Primary.ID
+		domain, err := tfec2.DescribeTransitGatewayMulticastDomain(conn, id)
+		if tfawserr.ErrMessageContains(err, "InvalidTransitGatewayMulticastDomainId.NotFound", "") {
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if domain == nil {
+			continue
+		}
+
+		state := aws.StringValue(domain.State)
+		if state != ec2.TransitGatewayMulticastDomainStateDeleted {
+			return fmt.Errorf(
+				"EC2 Transit Gateway Multicast Domain (%s) still exists in a non-deleted (%s) state",
+				id, state)
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckTransitGatewayMulticastDomainAssociations(domain *ec2.TransitGatewayMulticastDomain, count int, expected map[*ec2.TransitGatewayVpcAttachment][]*ec2.Subnet) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+		id := aws.StringValue(domain.TransitGatewayMulticastDomainId)
+
+		assocSet, err := tfec2.GetTransitGatewayMulticastDomainAssociations(conn, id)
+		if err != nil {
+			return err
+		}
+
+		assocLen := len(assocSet)
+		if assocLen != count {
```

Comment:
Assertions should not be done in the subroutines. Assertions should be done in the `testAcc` function. Use the subroutine to fetch the data and pass it by reference to the `testAcc`.